hal/shared/power_hal.cpp

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright (c) 2019 Particle Industries, Inc.  All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation, either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "power_hal.h"
+#include "dct.h"
+#include "check.h"
+#include "system_error.h"
+#include <algorithm>
+
+int hal_power_load_config(hal_power_config* conf, void* reserved) {
+    CHECK_TRUE(conf, SYSTEM_ERROR_INVALID_ARGUMENT);
+    int err = dct_read_app_data_copy(DCT_POWER_CONFIG_OFFSET, conf, conf->size);
```

Comment:
Nitpick: why not determine the conf size in `hal_power_load_config `, it's convenient that the user doesn't need to set the size. And it's safe for memory copy.

Reply:
For backwards compatibility. The stored configuration may potentially be larger (newer version) than what the caller can handle.

Reply:
That make sense, 👍